### PR TITLE
Enhancement: Append Existing Archive or Entry

### PR DESCRIPTION
**Pull request by guspower**
As per issue #177 here's a sample implementation with supporting tests and docs.

Let me know what you think.

### CODE REVIEW

File: src/builder.rs [hunk 1]
```diff
@@ -111,6 +111,73 @@ impl<W: Write> Builder<W> {
...
+    /// let mut input = Archive::new(File::open("/path/to/existing.tar").unwrap());
+    ///
+    /// let mut output_file = File::create("/path/to/target.tar").unwrap();
+    /// let mut output = Builder::new(output_file);
+    ///
+    /// output.append_archive(&mut input).unwrap();
+    /// output.finish().unwrap();
+    /// ```
+    pub fn append_archive<R: Read>(&mut self, archive: &mut Archive<R>) -> io::Result<()> {
+        let entries = archive.entries()?;
```
Comment (alexcrichton) [inline src/builder.rs]: I think this may want to use `entries_raw` perhaps? That way it'll handle entries-describing-other-entries I think

File: src/builder.rs [hunk 2]
```diff
@@ -111,6 +111,73 @@ impl<W: Write> Builder<W> {
...
+    /// let mut entry = input.entries().unwrap().next().unwrap().unwrap();
+    ///
+    /// let mut output_file = File::create("/path/to/target.tar").unwrap();
+    /// let mut output = Builder::new(output_file);
+    ///
+    /// output.append_entry(&mut entry).unwrap();
+    /// output.finish().unwrap();
+    /// ```
+    pub fn append_entry<R: Read>(&mut self, entry: &mut Entry<R>) -> io::Result<()> {
+        let header = entry.header().clone();
```
Comment (alexcrichton) [inline src/builder.rs]: I'm a little worried that this loses information because it doesn't handle things like long path names and such?

Comment (guspower) [reply src/builder.rs]: OK - can you explain this one a bit more? I see that in https://github.com/alexcrichton/tar-rs/blob/master/src/header.rs#L888 clone uses `self.bytes` - should I be checking `is_ustar` and `is_gnu` instead and constructing the appropriate `Header` instance?

Also could you describe an example that I could use as a test to verify this works as expected?

Cheers! G

Comment (alexcrichton) [reply src/builder.rs]: Oh sorry yeah, I should clarify! The `Entry<T>` type has a number of inherent methods:

* `Entry::path`
* `Entry::link_name`
* `Entry::path_extensions`

With various extensions to the `tar` format one instance of `Entry<T>` may actually be multiple tar-level entries in the underlying file. For example some entries just say "this is a longer path name than you could otherwise store". That's what the `raw(true)` for the entries iterator above is doing, it's avoiding automatically coalescing entries together and simply appending them one-by-one.

Here though an `Entry` may represent multiple tar-level entries that need to be emitted, but I believe the implementation here is only emitting one. Perhaps at least `path` could be handled? I think if you call `append_path` it should largely do the right thing.

Comment (guspower) [reply src/builder.rs]: Right I think I understand. I see looking at the [UStar format](https://en.wikipedia.org/wiki/Tar_(computing)#UStar_format) that the `'g'` and `'x'` header type flags can provide all sorts of stuff:
> extended header with meta data for the next file in the archive

I had no idea tar files could be that exciting!  

I'll add in some test cases that will hopefully surface the shortcoming in the implementation and then fix it up.

Comment (alexcrichton) [reply src/builder.rs]: Yeah tar files ended up being full of surprises, I thought writing a parser library like this wouldn't be too bad and here we are!

Comment (guspower) [reply src/builder.rs]: Yeah no kidding! 

I managed to replicate the issue with path truncation you mentioned, and have a simple change that handles long path names (as per your suggestion):
```self.append(&mut header, entry)?;``` -> ```self.append_data(&mut header, path, entry)?;```

The only issue I have is with a test assertion on the contents of the generated entry header - specifically with the checksum. It seems the `Builder` generates a different format checksum from `GNU tar`:

Builder | GNU tar
------------ | -------------
`0007666 `|`007666  `

i.e. the `Builder` creates a 7 digit representation followed by a space, while `GNU tar` creates a 6 digit representation followed by a NUL and then a space.

Wikipedia says:
> It is stored as a six digit octal number with leading zeroes followed by a NUL and then a space. Various implementations do not adhere to this format. For better compatibility, ignore leading and trailing whitespace, and take the first six digits. 

Do you think I should update the checksum implementation to generate `6 digits w/ leading zeros + NUL + space` so that it matches the `GNU tar` behaviour?

Comment (guspower) [reply src/builder.rs]: More specifically, I'm referring to the use of `octal_into` for checksum values https://github.com/alexcrichton/tar-rs/blob/master/src/header.rs#L1392 - I think the last two array values should be `NUL` and `space` (`[0, 32]`).

File: src/builder.rs [hunk 3]
```diff
@@ -111,6 +111,72 @@ impl<W: Write> Builder<W> {
...
+    /// let mut output = Builder::new(output_file);
+    ///
+    /// output.append_archive(&mut input).unwrap();
+    /// output.finish().unwrap();
+    /// ```
+    pub fn append_archive<R: Read>(&mut self, archive: &mut Archive<R>) -> io::Result<()> {
+        let entries = archive.entries()?.raw(true);
+
+        for entry in entries {
+            self.append_entry(entry?)?;
```
Comment (alexcrichton) [inline src/builder.rs]: I think this here, because it's a raw iterator, may actually want to use `append` directly for the reason that I think it's guaranteed to be correct

File: src/builder.rs [hunk 4]
```diff
@@ -111,6 +111,72 @@ impl<W: Write> Builder<W> {
...
+        }
+
+        Ok(())
+    }
+
+    /// Appends an entry from an existing archive to this archive.
+    ///
+    /// This function will append an existing `Entry` to the builder target.
```
Comment (alexcrichton) [inline src/builder.rs]: Could some caveats be mentioned here about things like PAX metadata and I think long symlink names not being preserved just yet?

File: src/header.rs [hunk 5]
```diff
@@ -677,7 +677,17 @@ impl Header {
     /// this header.
     pub fn set_cksum(&mut self) {
         let cksum = self.calculate_cksum();
-        octal_into(&mut self.as_old_mut().cksum, cksum);
+
+        let dst = &mut self.as_old_mut().cksum;
+        let o = format!("{:o}", cksum);
```
Comment (alexcrichton) [inline src/header.rs]: This has historically been really finnicky, but this change seems fine to me! Perhaps though the implementation could still use `octal_into`, but the slice passed in could be two elements shorter which are manually initialized here? (avoids the need to duplicate some of the implementation)

Comment (guspower) [reply src/header.rs]: yep I thought about pulling it all into `octal_into` but I thought it was a bit cryptic so I kept it separate (so you could more clearly see what was going on) - I'll update `octal_into` as you recommend.

File: src/builder.rs (hunk 4 above)
Comment (guspower) [reply src/builder.rs]: I'm struggling to replicate GNU `@LongLink` or PAX metadata loss in a test. I've added some additional test resources - `pax3.tar` and `gnu-longname.tar` - to try and push these edge cases (see [append_entry_from_existing_archive](https://github.com/alexcrichton/tar-rs/blob/0fe050da0a6702373b06d8192393ee2393f91187/tests/all.rs#L403)). 

Any ideas about how I can enhance these tests to catch the unsupported cases?

File: src/builder.rs (hunk 3 above)
Comment (guspower) [reply src/builder.rs]: Yep good point

File: .gitignore [hunk 6]
```diff
@@ -1,2 +1,7 @@
 /target
 /Cargo.lock
+
+#Intellij Rust IDE
+.idea
+*.iml
+*.ipr
```
Comment (dvc94ch) [inline .gitignore:7]: don't think these should be here. here is some more information on how to configure git https://gist.github.com/subfuzion/db7f57fff2fb6998a16c locally to ignore those files.

**Review by dvc94ch — APPROVED** [review]
Great, this looks like a very useful feature.

**Comment by guspower** [comment]
@dvc94ch If this PR is still considered useful I'm happy to bring it up-to-date.

**Comment by dvc94ch** [comment]
That would be great. Should be fairly easy, using this branch for now: https://github.com/dvc94ch/tar-rs/tree/append-entry.
